src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/UserTimeZoneService.cs

```diff
@@ -85,7 +89,8 @@ private async ValueTask<string> GetTimeZoneIdAsync(IUser user)
         // Do we know this user doesn't have a configured value?
         if (timeZoneId == EmptyTimeZone)
         {
-            return null;
+            // The user opted to use the site's time zone.
+            return (await _siteService.GetSiteSettingsAsync())?.TimeZoneId;
```

Comment:
I have not looked at the code to see how it is being consumed. But returning the site's timezone in a user timezone service sounds wrong to me. I think it should return null if the user does not have one. The consumer would then fall back on the size time zone if the is the business logic.

Reply:
@Piedone I did a Quick Look at the code and I agree with my own comment above. 

Here we should return null if `currentUser` is null or if `(await _userTimeZoneService.GetAsync(currentUser))?.TimeZoneId` is null. 


https://github.com/OrchardCMS/OrchardCore/blob/601651517ee811500d22a72ecbbdf0ab22e3d98b/src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/UserTimeZoneSelector.cs#L29


The UserTimeZoneService should not return the site's time zone since that will eventually be selected if there is no other selection with higher priority to select from.

Reply:
See my comment above: https://github.com/OrchardCMS/OrchardCore/pull/16777#discussion_r1769633190

Reply:
why is it a good design to not return a null as a result? If a user has no time zone, returning null would be valid.

Reply:
Because then we need to guard the method call with null checks, and forgetting it we get bugs like this one.

The user does have a time zone. `IUserTimeZoneService.GetAsync()` is about retrieving the time zone the user wants to use. We know it, and can retrieve it accordingly, even if ultimately from site settings. This is similar to any default value we use.

Reply:
but any time you return have an interface that return a nullable object, you should always check/guard against null if your code can't handle it.

Reply:
I can understand both of your arguments, but in this case I think the actual solution is to check the site settings in the `UserTimeZoneSelector` instead. This makes more sense to me and actually satisfies both of your arguments :)

Reply:
While I disagree due to the above-mentioned arguments (and the existence of this very bug), I bow before the majority. You guys look for the next such NRE, though ;).

Reply:
@Piedone you can include a test case to ensure we don't run into such an issue

Reply:
This is about consumers of `IUserTimeZoneService`. There's no test case we can add for that.